lib/_http_agent.js

```diff
@@ -116,10 +117,29 @@ function Agent(options) {
       }
     }
   });
+
+  // Don't emit keylog events unless there is a listener for them.
+  this.on('newListener', maybeEnableKeylog);
 }
 Object.setPrototypeOf(Agent.prototype, EventEmitter.prototype);
 Object.setPrototypeOf(Agent, EventEmitter);
 
+function maybeEnableKeylog(eventName) {
+  if (eventName === 'keylog') {
+    this.removeListener('newListener', maybeEnableKeylog);
+    // Future sockets will listen on keylog at creation.
+    const agent = this;
+    this[kOnKeylog] = function onkeylog(keylog) {
+      agent.emit('keylog', keylog, this);
```

Comment:
`this === agent` here, isn't it?

Reply:
No, TLSSocket is using it as a listener later with something like `tlsSock.on('keylog', agent[kOnKeylog])`, so `this` is the TLSSocket here and `agent` has to be in the closure so the event can be emitted on it. The test assert it this (well, duck type, they assert that it has the `.encrypted` property, which agent and net.Socket don't).